### PR TITLE
consensus: HasVoteMessage index boundary check

### PR DESCRIPTION
**Pull request by JayT106**
Went through #2871, there are several issues, this PR tries to tackle the `HasVoteMessage` with an invalid validator index sent by a bad peer and it prevents the bad vote goes to the peerMsgQueue.

Future work, check other bad message cases and plumbing the reactor errors with the peer manager and then can disconnect the peer sending the bad messages.

### CODE REVIEW

File: internal/consensus/peer_state.go [hunk 1]
```diff
@@ -19,6 +19,8 @@ import (
 var (
 	ErrPeerStateHeightRegression = errors.New("peer state height regression")
 	ErrPeerStateInvalidStartTime = errors.New("peer state invalid startTime")
+	ErrPeerStateNilVote          = errors.New("peer state nil vote")
+	ErrPeerStateInvalidVoteIndex = errors.New("peer state invalid vote index")
```
Comment (cmwaters) [inline internal/consensus/peer_state.go:23]: Could we perhaps have a more descriptive error message i.e. `peer sent a vote with an invalid index`

File: internal/consensus/reactor.go [hunk 2]
```diff
@@ -670,7 +670,11 @@ func (r *Reactor) pickSendVote(ctx context.Context, ps *PeerState, votes types.V
 		return false, err
 	}
 
-	ps.SetHasVote(vote)
+	// the vote should be non-nil.
```
[the remark this answers is not in the document]
Comment (JayT106) [reply internal/consensus/reactor.go:673]: you are right! I will remove this comment.

File: internal/consensus/peer_state.go (hunk 1 above)
Comment (JayT106) [reply internal/consensus/peer_state.go]: yes, updated!

File: internal/consensus/reactor.go [hunk 3]
```diff
@@ -1101,8 +1104,10 @@ func (r *Reactor) handleStateMessage(ctx context.Context, envelope *p2p.Envelope
 		ps.ApplyNewValidBlockMessage(msgI.(*NewValidBlockMessage))
 
 	case *tmcons.HasVote:
-		ps.ApplyHasVoteMessage(msgI.(*HasVoteMessage))
-
+		if err := ps.ApplyHasVoteMessage(msgI.(*HasVoteMessage)); err != nil {
+			r.logger.Error("peer sent us an invalid HasVote msg", "msg", msg, "err", err)
```
Comment (williambanfield) [inline internal/consensus/reactor.go:1108]:
```suggestion
			r.logger.Error("applying HasVote message", "msg", msg, "err", err)
```
The `err` value contains the specific info here.

File: libs/bits/bit_array.go [hunk 4]
```diff
@@ -72,7 +72,7 @@ func (bA *BitArray) getIndex(i int) bool {
 }
 
 // SetIndex sets the bit at index i within the bit array.
-// The behavior is undefined if i >= bA.Bits
+// The behavior is undefined if i >= bA.Bits or i < 0
```
Comment (williambanfield) [inline libs/bits/bit_array.go:75]:
```suggestion
// This method returns false if i is out of range of the BitArray. 
```
I think this is a slightly more accurate comment.

**Review by cmwaters — APPROVED** [review]
👍

**Pull request merged**